README.md
TPs de Paradigmas de Lenguajes de Programación (FCEyN, UBA).

Realizados durante el 2do cuatrimestre de 2017.


tp1/changelog.md
Cambios
-------

* Usar case y where en `foldA23`
* Agregar comentario sobre composición de concats en `internos` y `hojas`
* Sacar argumentos innecesarios en mapA23
* Refactorizar `truncar` para hacerlo mas entendible
* Cambiar la comparación con [] por un `null` en `obtener`
* Test del ej7 más relacionado a la función `definir`

Dudas
-----

* La corrección dice que los tests no compilan,
    pero nos funcionan bien con GHC 8.0.2 y HUnit 1.6.0.0


tp1/Main.hs
import Diccionario
import Data.List
import Data.Maybe
import Arbol23
import Test.HUnit

--Este módulo sirve para utilizar el diccionario sin acceder a su estructura interna. Pueden agregar otras funciones o casos de prueba.

{- Función a implementar. -}

búsquedaDelTesoro :: Eq a => a -> (a -> Bool) -> Diccionario a a -> Maybe a
búsquedaDelTesoro x esTesoro dicc = head $ dropWhile niTesoroNiVacio (iterate f $ Just x)
  where f x = obtener (fromJust x) dicc
        niTesoroNiVacio x = isJust x && not (esTesoro $ fromJust x)

{- Diccionarios de prueba: -}

dicc1 :: Diccionario Int String
dicc1 = definirVarias [(0,"Hola"),(-10,"Chau"),(15,"Felicidades"),(2,"etc."),(9,"a")]
                      (vacio (<))

dicc2 :: Diccionario String String
dicc2 = definirVarias [("inicio","casa"),("auto","flores"),("calle","auto"),
                       ("casa","escalera"),("ropero","alfajor"),("escalera","ropero")]
                       (vacio (<))

dicc3 :: Diccionario Int String
dicc3 = definirVarias [(0,"Hola"),(-10,"Chau"),(15,"Felicidades"),(2,"etc."),(9,"a")]
                      (vacio (\x y->x `mod` 5 < y `mod` 5))

--Ejecución de los tests
main :: IO Counts
main = do runTestTT allTests

allTests = test [
    "ejercicio2" ~: testsEj2,
    "ejercicio3" ~: testsEj3,
    "ejercicio4" ~: testsEj4,
    "ejercicio5" ~: testsEj5,
    "ejercicio6" ~: testsEj6,
    "ejercicio7" ~: testsEj7,
    "ejercicio8" ~: testsEj8,
    "ejercicio9" ~: testsEj9,
    "ejercicio10" ~: testsEj10
  ]

testsEj2 = test [
    [0,1,2,3,4,5,6,7] ~=? internos arbolito1,
    "abcdefghi" ~=? hojas arbolito1,
    [True,False,True] ~=? internos arbolito2,
    [1,2,3,2,3,4,3,4,5,4] ~=? take 10 (hojas arbolito3)
  ]

testsEj3 = test [
    [0,1,-1,5] ~=? hojas (incrementarHojas arbolito2)
  ]

testsEj4 = test [
    [1,2,3,2,3,4,3,4,5,4,5,6,0,0,0,0,0] ~=? hojas (truncar 0 6 arbolito3)
  ]

testsEj5 = test [
    22 ~=? evaluar (truncar 0 6 arbolito3)
  ]

testsEj6 = test [
    True ~=? (isNothing . estructura) (vacio (<) :: Diccionario Int Int)
  ]

testsEj7 = test [
     0 ~=? (length . concat . fmap hojas . estructura) (vacio (<) :: Diccionario Int Int),

     -- dicc1 se crea con definirVarias que a su vez llama a definir
     [(-10,"Chau"),(0,"Hola"),(2,"etc."),(9,"a"),(15,"Felicidades")] ~=? (concat . fmap hojas . estructura) dicc1,
     [(-10,"Chau"),(0,"Hola"),(1,"Hi"),(2,"etc."),(9,"a"),(15,"Felicidades")] ~=? (concat . fmap hojas . estructura . definir 1 "Hi") dicc1
  ]

testsEj8 = test [
    Nothing ~=? obtener 42 (vacio (<) :: Diccionario Int Int),

    Nothing ~=? obtener 1 dicc1,
    Just "a" ~=? obtener 9 dicc1,
    Just "Hi" ~=? (obtener 1 . definir 1 "Hi") dicc1,

    Just "auto" ~=? obtener "calle" dicc2,

    Nothing ~=? obtener 4 dicc3,
    Just "a" ~=? obtener 9 dicc3,
    Just "Hi" ~=? (obtener 4 . definir 4 "Hi") dicc3
  ]

testsEj9 = test [
    [] ~=? claves (vacio (<) :: Diccionario Int Int),

    sort [0, -10, 15, 2, 9] ~=? (sort . claves) dicc1,
    sort [0, -10, 15, 2, 9, 1] ~=? (sort . claves . definir 1 "Hi") dicc1,

    sort ["inicio", "auto", "calle", "casa", "ropero", "escalera"]
            ~=? (sort . claves) dicc2
  ]

testsEj10 = test [
    Just "alfajor" ~=? búsquedaDelTesoro "inicio" ((=='a').head) dicc2
  ]



tp1/Arbol23.hs
module Arbol23 where

data Arbol23 a b = Hoja a | Dos b (Arbol23 a b) (Arbol23 a b)
                 | Tres b b (Arbol23 a b) (Arbol23 a b) (Arbol23 a b)

{- Funciones para mostrar el árbol. -}

instance (Show a, Show b) => Show (Arbol23 a b) where
    show = ("\n" ++) . (padTree 0 0 False)

padlength = 5

padTree :: (Show a, Show b) => Int -> Int -> Bool -> (Arbol23 a b) -> String
padTree nivel acum doPad t = case t of
        (Hoja x) -> initialPad ++ stuff x
        (Dos x i d) -> initialPad ++ stuff x ++
                pad padlength ++ rec x False i ++ "\n" ++
                rec x True d ++ "\n"
        (Tres x y i m d) -> initialPad ++ stuff x ++ --(' ':tail (stuff y)) ++
                pad padlength ++ rec x False i ++ "\n" ++
                pad levelPad ++ stuff y ++ pad padlength ++ rec x False m ++ "\n" ++
                rec x True d ++ "\n"
    where l = length . stuff
          levelPad = (padlength*nivel + acum)
          initialPad = (if doPad then pad levelPad else "")
          rec x = padTree (nivel+1) (acum+l x)

stuff :: Show a => a -> String
stuff x = if n > l then pad (n-l) ++ s else s
    where s = show x
          l = length s
          n = padlength

pad :: Int -> String
pad i = replicate i ' '

{- Funciones pedidas. -}

foldA23 :: (a->c) -> (b->c->c->c) -> (b->b->c->c->c->c) -> Arbol23 a b -> c
foldA23 f g h a = case a of
                    (Hoja x) -> f x
                    (Dos x t1 t2) -> g x (fld t1) (fld t2)
                    (Tres x y t1 t2 t3) -> h x y (fld t1) (fld t2) (fld t3)
    where fld = foldA23 f g h

--Lista en preorden de los internos del árbol.
    --
    -- Componemos funciones que construyen listas en vez de concatenarlas
    -- explícitamente pues la concatenación tarda tiempo lineal en el tamaño
    -- del primer argumento (tiene que reconstruir la lista),
    -- por lo que `internos` tardaría tiempo cuadrático en la cantidad de nodos
    --
    -- En cambio si componemos funciones y recién las evaluamos al final
    -- podemos hacer todo el proceso en tiempo lineal
internos :: Arbol23 a b -> [b]
internos arbol = foldA23 (const id)
                         (\x r1 r2 -> (x:) . r1 . r2)
                         (\x y r1 r2 r3 -> (x:) . (y:) . r1 . r2 . r3) arbol []

--Lista las hojas de izquierda a derecha.
    -- Ver el comentario de `internos`
hojas :: Arbol23 a b -> [a]
hojas a = foldA23 (:) (\x r1 r2 -> r1 . r2) (\x y r1 r2 r3 -> r1 . r2 . r3) a []

esHoja :: Arbol23 a b -> Bool
esHoja = foldA23 (const True) (\x r1 r2 -> False) (\x y r1 r2 r3 -> False)

mapA23 :: (a -> c) -> (b -> d) -> Arbol23 a b -> Arbol23 c d
mapA23 fa fb = foldA23 (Hoja . fa)
                       (Dos . fb)
                       (\x y -> Tres (fb x) (fb y))


--Ejemplo de uso de mapA23.
--Incrementa en 1 el valor de las hojas.
incrementarHojas :: Num a => Arbol23 a b -> Arbol23 a b
incrementarHojas = mapA23 (+1) id


--Trunca el árbol hasta un determinado nivel. Cuando llega a 0, reemplaza el resto del árbol por una hoja con el valor indicado.
--Funciona para árboles infinitos.
truncar :: a -> Integer -> Arbol23 a b -> Arbol23 a b
truncar h n t = foldA23 fnHoja fnDos fnTres t n
    where cutAt i t = if i == 0 then Hoja h else t
          fnHoja x i = cutAt i $ Hoja x
          fnDos x r1 r2 i = cutAt i $ Dos x (r1 $ i-1) (r2 $ i-1)
          fnTres x y r1 r2 r3 i = cutAt i $ Tres x y (r1 $ i-1) (r2 $ i-1) (r3 $ i-1)

--Evalúa las funciones tomando los valores de los hijos como argumentos.
--En el caso de que haya 3 hijos, asocia a izquierda.
evaluar :: Arbol23 a (a->a->a) -> a
evaluar = foldA23 id id (\f g t1 t2 t3 -> g (f t1 t2) t3)

--Ejemplo:
--evaluar (truncar 0 6 arbolito3) = 22 = (1*2-3)+(2*3-4)+(3*4-5)+(4*5-6)

{- Árboles de ejemplo. -}
arbolito1 :: Arbol23 Char Int
arbolito1 = Tres 0 1
            (Dos 2 (Hoja 'a') (Hoja 'b'))
            (Tres 3 4 (Hoja 'c') (Hoja 'd') (Dos 5 (Hoja 'e') (Hoja 'f')))
            (Dos 6 (Hoja 'g') (Dos 7 (Hoja 'h') (Hoja 'i')))

arbolito2 :: Arbol23 Int Bool
arbolito2 = Dos True (Hoja (-1)) (Tres False True (Hoja 0) (Hoja (-2)) (Hoja 4))

arbolito3 :: Arbol23 Int (Int->Int->Int)
arbolito3 = Dos (+) (Tres (*) (-) (Hoja 1) (Hoja 2) (Hoja 3)) (incrementarHojas arbolito3)

arbolito4 :: Arbol23 Int Char
arbolito4 = Dos 'p' (Dos 'l' (Dos 'g' (Hoja 5) (Hoja 2)) (Tres 'r' 'a' (Hoja 0)(Hoja 1)(Hoja 12)))
                    (Dos 'p' (Tres 'n' 'd' (Hoja (-3))(Hoja 4)(Hoja 9)) (Dos 'e' (Hoja 20)(Hoja 7)))



tp2/esqueleto.pl
%------------------Predicados predefinidos:------------------%

%fliplength(?Longitud, ?Lista)
fliplength(N, L) :- length(L, N).

%matriz(?Matriz, ?Filas, ?Columnas)
matriz(M, F, C) :- length(M, F), maplist(fliplength(C), M).

%dif1(+N1, ?N2)
dif1(N1, N2) :- N2 is N1 + 1.
dif1(N1, N2) :- N2 is N1 - 1.

%adyacente(+F1, +C1, ?F2, ?C2)
adyacente(F1,C1,F1,C2) :- dif1(C1,C2).
adyacente(F1,C1,F2,C1) :- dif1(F1,F2).
adyacente(F1,C1,F2,C2) :- dif1(C1,C2), dif1(F1,F2).

%enRango(+Matriz, +Fila, +Columna)
enRango([Fila|Filas], F, C) :- F > 0, C > 0, length([Fila|Filas], FMax), F =< FMax, length(Fila, CMax), C =< CMax.

%adyacenteEnRango(+Tablero, +F1, +C1, ?F2, ?C2)
adyacenteEnRango(T,F1,C1,F2,C2) :- adyacente(F1,C1,F2,C2), enRango(T,F2,C2).

%------------------Predicados extra:------------------%

%free(+Tablero, +Fila, +Columna)
free(Tablero, Fila, Columna) :- nth1(Fila, Tablero, Row), nth1(Columna, Row, X), var(X).

%instanciarCasillero(?X)
instanciarCasillero(X) :- var(X), X = ~ .
instanciarCasillero(X) :- nonvar(X).

%------------------Predicados a definir:------------------%

%contenido(+?Tablero, ?Fila, ?Columna, ?Contenido)

%disponible(+Tablero, ?Fila, ?Columna)
disponible(Tablero, Fila, Columna) :- free(Tablero, Fila, Columna),
    forall(adyacenteEnRango(Tablero, Fila, Columna, F, C), free(Tablero, F, C)).

%puedoColocar(+CantPiezas, ?Direccion, +Tablero, ?Fila, ?Columna)

%ubicarBarcos(+Barcos, +?Tablero)

%completarConAgua(+?Tablero)
completarConAgua(Tablero) :- maplist(maplist(instanciarCasillero), Tablero).

%golpear(+Tablero, +NumFila, +NumColumna, -NuevoTab)

% Completar instanciación soportada y justificar.
%atacar(Tablero, Fila, Columna, Resultado, NuevoTab)

%------------------Tests:------------------%

test(1) :- matriz(M,2,3), adyacenteEnRango(M,2,2,2,3).
test(2) :- matriz(M,2,3), setof((F,C), adyacenteEnRango(M,1,1,F,C), [ (1, 2), (2, 1), (2, 2)]).

test(3) :- T = [[~, _, _],[o, _, _]], contenido(T,1,1,~), contenido(T,2,1,o).
test(4) :- T = [[~, _, _],[o, _, _]], contenido(T,1,1,X), contenido(T,2,1,Y), X = ~, Y = o.
test(5) :- T = [[o, ~],[o, ~]], setof((Y,X,C), contenido(T,Y,X,C), [(1,1,o), (1,2,~), (2,1,o), (2,2,~)]).

test(6) :- T = [[o, _, _],[~, _, _]], disponible(T,2,3), not(disponible(T,1,2)).
test(7) :- T = [[o, _, _],[~, _, _]], setof((Y,X), disponible(T,Y,X), [(1,3), (2,3)]).

test(8) :- matriz(M,2,3), contenido(M,2,1,o), puedoColocar(2,vertical,M,1,3).
test(9) :- matriz(M,2,3), setof((Dir,F,C), puedoColocar(3,Dir,M,F,C),
                              [(horizontal,1,1), (horizontal,1,2), (horizontal,2,1), (horizontal,2,2)]).
test(10) :- matriz(M,2,2), setof((Dir,F,C), puedoColocar(2,Dir,M,F,C),
                              [(horizontal,1,1), (horizontal,2,1), (vertical,1,1), (vertical,1,2)]).

test(11) :- matriz(M,3,2), setof(M, ubicarBarcos([2,1],M), [ [[o,o], [_,_], [o,_]], [[o,o], [_,_], [_,o]],
                                                                 [[o,_], [_,_], [o,o]], [[_,o], [_,_], [o,o]] ]).

test(12) :- matriz(M,3,2), completarConAgua(M), not(free(M,_,_)).
test(13) :- T = [[o, _, _],[~, _, o]], completarConAgua(T), T = [[o, ~, ~],[~, ~, o]].

test(14) :- T = [[o, _, _],[~, _, o]], golpear(T,2,3,R), R = [[o, _, _],[~, _, ~]].
test(15) :- T = [[o, _, _],[~, _, o]], golpear(T,2,2,R), R = [[o, _, _],[~, ~, o]].

test(16) :- T = [[o,o], [~,~], [~,o]], atacar(T,1,1,Res,R), Res = tocado, R = [[~,o], [~,~], [~,o]].
test(17) :- T = [[o,o], [~,~], [~,o]], atacar(T,3,1,Res,R), Res = agua, R = T.
test(18) :- T = [[o,o], [~,~], [~,o]], atacar(T,3,2,Res,R), Res = hundido, R = [[~,o], [~,~], [~,~]].

tests :- forall(between(1,18,N), test(N)). % Cambiar el 2 por la cantidad de tests que tengan.



tp1/Diccionario.hs
-- Exportamos estructura para poder testear vacio
module Diccionario (Diccionario(estructura), vacio, definir, definirVarias, obtener, claves) where

import Prelude hiding (concat)
import Data.Foldable (concat)
import Data.Maybe
import Data.List
import Arbol23

{- Definiciones de tipos. -}

type Comp clave = clave->clave->Bool
type Estr clave valor = Arbol23 (clave,valor) clave

data Diccionario clave valor = Dicc {cmp :: Comp clave, estructura :: Maybe (Estr clave valor)}
--El comparador es por menor.

{- Funciones provistas por la cátedra. -}

--Inserta un nuevo par clave, valor en una estructura que ya tiene al menos un dato.
insertar :: clave -> valor -> Comp clave -> Estr clave valor -> Estr clave valor
insertar c v comp a23 = interceptar (insertarYPropagar c v comp a23) id (\s1 (c1, s2)->Dos c1 s1 s2)

--Maneja el caso de que la segunda componente sea Nothing.
interceptar :: (a, Maybe b) -> (a -> c) -> (a -> b -> c) -> c
interceptar (x,y) f1 f2 = case y of
                   Nothing -> f1 x
                   Just z -> f2 x z

{- Inserta una clave con su valor correspondiente. Si se actualiza el índice, el cambio se propaga hacia arriba
   para mantener balanceado el árbol.
   Usamos recursión explícita porque este tipo de recursión no es estructural (no se aplica a todos los hijos). -}
insertarYPropagar :: clave -> valor -> Comp clave -> Estr clave valor ->
                     (Estr clave valor, Maybe (clave, Estr clave valor))
insertarYPropagar c v comp a23 =
    let rec = insertarYPropagar c v comp in
        case a23 of
        --Si es hoja, elegimos la máxima de las claves y propagamos el balanceo hacia arriba.
            Hoja (ch,vh) -> if comp c ch 
                then (Hoja (c,v), Just (ch, Hoja (ch,vh)))
                else (Hoja (ch, vh), Just (c, Hoja (c,v)))
        {- Si el actual es Nodo-Dos, o se queda en forma Nodo-Dos o se transforma en 
        Nodo-Tres; no puede ocurrir que haya propagación hacia arriba (retornamos Nothing). -}
            Dos c1 a1 a2 -> (if comp c c1
                then
                -- La clave c va del lado izquierdo.
                    interceptar (rec a1) 
                        (\s1 -> Dos c1 s1 a2)
                        (\s1 (c3, s2) -> Tres c3 c1 s1 s2 a2)
                else
                -- La clave c va del lado derecho.
                    interceptar (rec a2) 
                        (\s1 -> Dos c1 a1 s1)
                        (\s1 (c3, s2) -> Tres c1 c3 a1 s1 s2), Nothing)
        {- Nodo-tres sólo propaga si de abajo propagan, los tres casos son muy similares
        Sólo cambia en que árbol se inserta. -}
            Tres c1 c2 a1 a2 a3 -> if comp c c1
                then
                    -- La clave debe ir en el primer árbol.
                    interceptar (rec a1) 
                        (\s1 -> (Tres c1 c2 s1 a2 a3, Nothing))
                        (\s1 (c3, s2) -> (Dos c3 s1 s2, Just(c1, Dos c2 a2 a3)))
                else if comp c c2
                    then
                        -- La clave debe ir en el árbol del medio.
                        interceptar (rec a2) 
                            (\s1 -> (Tres c1 c2 a1 s1 a3, Nothing))
                            (\s1 (c3, s2) -> (Dos c1 a1 s1, Just(c3, Dos c2 s2 a3)))
                    else
                        --La clave debe ir en el último árbol.
                        interceptar (rec a3) 
                            (\s1 -> (Tres c1 c2 a1 a2 s1, Nothing))
                            (\s1 (c3, s2) -> (Dos c1 a1 a2, Just(c2, Dos c3 s1 s2)))

--Se asume que la lista no tiene claves repetidas.
definirVarias :: [(clave,valor)] -> Diccionario clave valor -> Diccionario clave valor
definirVarias = (flip . foldr . uncurry) definir

{- Funciones a implementar. -}

vacio :: Comp clave -> Diccionario clave valor
vacio c = Dicc c Nothing

definir :: clave -> valor -> Diccionario clave valor -> Diccionario clave valor
definir k v dicc = Dicc comp nuevaEstr 

  where comp = cmp dicc
        nuevaEstr = if isNothing(estructura dicc)
                    then Just (Hoja (k,v))
                    else Just (insertar k v comp a23)
        a23 = fromJust (estructura dicc)

obtener :: Eq clave => clave -> Diccionario clave valor -> Maybe valor
obtener k dicc = if null $ claves dicc then Nothing else foldA23 casoHoja casoDos casoTres a23
  where casoHoja = \h -> if k == fst(h) then Just (snd h) else Nothing
        casoDos = \x t1 t2 -> if comp k x then t1 else t2
        casoTres = \x y t1 t2 t3 -> if comp k x then t1 else (if comp k y then t2 else t3)
        comp = cmp dicc
        a23 = fromJust (estructura dicc)

claves :: Diccionario clave valor -> [clave]
claves = concat . fmap (map fst . hojas) . estructura

{- Diccionarios de prueba: -}

dicc1 :: Diccionario Int String
dicc1 = definirVarias [(0,"Hola"),(-10,"Chau"),(15,"Felicidades"),(2,"etc."),(9,"a")]
                      (vacio (<))

dicc2 :: Diccionario String String
dicc2 = definirVarias [("inicio","casa"),("auto","flores"),("calle","auto"),
                       ("casa","escalera"),("ropero","alfajor"),("escalera","ropero")]
                      (vacio (<))

dicc3 :: Diccionario Int String
dicc3 = definirVarias [(0,"Hola"),(-10,"Chau"),(15,"Felicidades"),(2,"etc."),(9,"a")]
                      (vacio (\x y->x `mod` 5 < y `mod` 5))



tp2/tp2.pl
%------------------Predicados predefinidos:------------------%

%fliplength(?Longitud, ?Lista)
fliplength(N, L) :- length(L, N).

%matriz(?Matriz, ?Filas, ?Columnas)
matriz(M, F, C) :- length(M, F), maplist(fliplength(C), M).

%dif1(+N1, ?N2)
dif1(N1, N2) :- N2 is N1 + 1.
dif1(N1, N2) :- N2 is N1 - 1.

%adyacente(+F1, +C1, ?F2, ?C2)
adyacente(F1,C1,F1,C2) :- dif1(C1,C2).
adyacente(F1,C1,F2,C1) :- dif1(F1,F2).
adyacente(F1,C1,F2,C2) :- dif1(C1,C2), dif1(F1,F2).

%enRango(+Matriz, +Fila, +Columna)
enRango([Fila|Filas], F, C) :- F > 0, C > 0, length([Fila|Filas], FMax), F =< FMax, length(Fila, CMax), C =< CMax.

%adyacenteEnRango(+Tablero, +F1, +C1, ?F2, ?C2)
adyacenteEnRango(T,F1,C1,F2,C2) :- adyacente(F1,C1,F2,C2), enRango(T,F2,C2).

%------------------Predicados extra:------------------%

%free(+Tablero, +Fila, +Columna)
free(Tablero, Fila, Columna) :- contenido(Tablero, Fila, Columna, X), var(X).

%proximaPosicion(?Direccion, +F1, +C1, -F2, -C2)
proximaPosicion(horizontal, F1, C1, F1, C2) :- succ(C1, C2).
proximaPosicion(vertical, F1, C1, F2, C1) :- succ(F1, F2).

%ubicarUnBarco(+Barco, ?Direccion, +?Tablero, +Fila, +Columna)
ubicarUnBarco(1, _, Tablero, Fila, Columna) :- contenido(Tablero, Fila, Columna, o).
ubicarUnBarco(Barco, Direccion, Tablero, Fila, Columna) :-
    M is Barco-1, proximaPosicion(Direccion, Fila, Columna, F, C),
    contenido(Tablero, Fila, Columna, o), ubicarUnBarco(M, Direccion, Tablero, F, C).

%instanciarCasillero(?X)
instanciarCasillero(X) :- var(X), X = ~ .
instanciarCasillero(X) :- nonvar(X).

%------------------Predicados a definir:------------------%

%contenido(+?Tablero, ?Fila, ?Columna, ?Contenido)
contenido(Tablero, Fila, Columna, Contenido) :-
    nth1(Fila, Tablero, Row), nth1(Columna, Row, X), X = Contenido.

%disponible(+Tablero, ?Fila, ?Columna)
disponible(Tablero, Fila, Columna) :- free(Tablero, Fila, Columna),
    forall(adyacenteEnRango(Tablero, Fila, Columna, F, C), free(Tablero, F, C)).

%puedoColocar(+CantPiezas, ?Direccion, +Tablero, ?Fila, ?Columna)
puedoColocar(1, _, Tablero, Fila, Columna) :- disponible(Tablero, Fila, Columna).
puedoColocar(CantPiezas, Direccion, Tablero, Fila, Columna) :- 
    contenido(Tablero, Fila, Columna, _),
    CantPiezas > 1, CantPiezasPred is CantPiezas - 1,
    proximaPosicion(Direccion, Fila, Columna, ProximaFila, ProximaColumna),
    disponible(Tablero, Fila, Columna),
    puedoColocar(CantPiezasPred, Direccion, Tablero, ProximaFila , ProximaColumna).

%ubicarBarcos(+Barcos, +?Tablero)
ubicarBarcos([], _).
ubicarBarcos([Barco|Barcos], Tablero) :- puedoColocar(Barco, D, Tablero, F, C),
                     ubicarUnBarco(Barco, D, Tablero, F, C),
                     ubicarBarcos(Barcos, Tablero).

%completarConAgua(+?Tablero)
completarConAgua(Tablero) :- maplist(maplist(instanciarCasillero), Tablero).

%golpear(+Tablero, ?NumFila, ?NumColumna, ?NuevoTab)
golpear(Tablero, NumFila, NumColumna, NuevoTab) :-
    matriz(Tablero, CantFilas, CantColumnas),
    matriz(NuevoTab, CantFilas, CantColumnas),
    contenido(NuevoTab, NumFila, NumColumna, ~),
    foreach((contenido(Tablero,I,J,MismoContenido), not((I =:= NumFila, J =:= NumColumna))),
           contenido(NuevoTab, I, J, MismoContenido)).

% Completar instanciación soportada y justificar.
%atacar(+Tablero, ?Fila, ?Columna, ?Resultado, ?NuevoTab)
%
% Notar que cambiamos el modo de los parámetros de atacar y golpear,
% indicando que son reversibles en todos sus parámetros excepto el Tablero inicial
%
% Tablero debe estar siempre instanciado pues así lo requiere matriz dentro de golpear
% Filas y Columna se generan a partir del predicado nth1(?Index, ?List, ?Elem)
% El Resultado solo puede estar en {agua, hundido, tocado}
% El NuevoTab se puede fijar y de esta forma filtrar las soluciones posibles, dado que golpear se encarga de instanciar su contenido. NuevoTab se instancia respecto del contenido ya instanciado del Tablero original.

atacar(Tablero, Fila, Columna, agua, Tablero) :- golpear(Tablero, Fila, Columna, Tablero).

atacar(Tablero, Fila, Columna, hundido, NuevoTab) :-
    contenido(Tablero, Fila, Columna, o),
    golpear(Tablero, Fila, Columna, NuevoTab),
    foreach(adyacenteEnRango(Tablero, Fila, Columna, F, C), not(contenido(Tablero, F, C, o))).

atacar(Tablero, Fila, Columna, tocado, NuevoTab) :-
    contenido(Tablero, Fila, Columna, o),
    golpear(Tablero, Fila, Columna, NuevoTab),
    adyacenteEnRango(Tablero, Fila, Columna, F, C),
    contenido(Tablero, F, C, o).

%------------------Tests:------------------%

test(1) :- matriz(M,2,3), adyacenteEnRango(M,2,2,2,3).
test(2) :- matriz(M,2,3), setof((F,C), adyacenteEnRango(M,1,1,F,C), [ (1, 2), (2, 1), (2, 2)]).

test(3) :- T = [[~, _, _],[o, _, _]], contenido(T,1,1,~), contenido(T,2,1,o).
test(4) :- T = [[~, _, _],[o, _, _]], contenido(T,1,1,X), contenido(T,2,1,Y), X = ~, Y = o.
test(5) :- T = [[o, ~],[o, ~]], setof((Y,X,C), contenido(T,Y,X,C), [(1,1,o), (1,2,~), (2,1,o), (2,2,~)]).

test(6) :- T = [[o, _, _],[~, _, _]], disponible(T,2,3), not(disponible(T,1,2)).
test(7) :- T = [[o, _, _],[~, _, _]], setof((Y,X), disponible(T,Y,X), [(1,3), (2,3)]).

test(8) :- matriz(M,2,3), contenido(M,2,1,o), puedoColocar(2,vertical,M,1,3).
test(9) :- matriz(M,2,3), setof((Dir,F,C), puedoColocar(3,Dir,M,F,C),
                              [(horizontal,1,1), (horizontal,2,1)]).
test(10) :- matriz(M,2,2), setof((Dir,F,C), puedoColocar(2,Dir,M,F,C),
                              [(horizontal,1,1), (horizontal,2,1), (vertical,1,1), (vertical,1,2)]).

test(11) :- matriz(M,3,2), setof(M, ubicarBarcos([2,1],M), [ [[o,o], [_,_], [o,_]], [[o,o], [_,_], [_,o]],
                                                                 [[o,_], [_,_], [o,o]], [[_,o], [_,_], [o,o]] ]).

test(12) :- matriz(M,3,2), completarConAgua(M), not(free(M,_,_)).
test(13) :- T = [[o, _, _],[~, _, o]], completarConAgua(T), T = [[o, ~, ~],[~, ~, o]].

test(14) :- T = [[o, _, _],[~, _, o]], golpear(T,2,3,R), R = [[o, _, _],[~, _, ~]].
test(15) :- T = [[o, _, _],[~, _, o]], golpear(T,2,2,R), R = [[o, _, _],[~, ~, o]].

test(16) :- T = [[o,o], [~,~], [~,o]], atacar(T,1,1,Res,R), Res = tocado, R = [[~,o], [~,~], [~,o]].
test(17) :- T = [[o,o], [~,~], [~,o]], atacar(T,3,1,Res,R), Res = agua, R = T.
test(18) :- T = [[o,o], [~,~], [~,o]], atacar(T,3,2,Res,R), Res = hundido, R = [[o,o], [~,~], [~,~]].

% reversibilidad en filas y columnas (en los anteriores se puede ver reversibilidad en Res y R, dado que no están instanciados antes de atacar)
test(19) :- T = [[o,o], [~,~], [~,o]], atacar(T,F,C,Res,R), Res = hundido, R = [[o,o], [~,~], [~,~]], F = 3, C = 2.
test(20) :- T = [[o,o], [~,~], [~,o]], atacar(T,F,C,Res,R), Res = hundido, R = [[o,o], [~,~], [~,~]], C = 2, F = 3.

tests(N) :- forall(between(1,N,I), test(I)).
tests :- tests(20).

